File: taurus-protect-sdk-typescript/src/mappers/balance.ts
/**
 * Balance mapper functions for converting OpenAPI DTOs to domain models.
 */

import type { TgvalidatordAssetBalance } from '../internal/openapi/models/TgvalidatordAssetBalance';
import type { TgvalidatordNFTCollectionBalance } from '../internal/openapi/models/TgvalidatordNFTCollectionBalance';
import type { AssetBalance, NFTCollectionBalance } from '../models/balance';
import { safeInt, safeMap } from './base';

/**
 * Maps a balance row to an AssetBalance domain model.
 *
 * The row is `{asset, balance}`: the currency lives in `asset.currencyInfo` and the amounts
 * in the nested `balance` object. The flat `currency` / `balance` keys this mapper used to
 * read do not exist on the wire, so every row came back without a currency and with a
 * `"[object Object]"` balance.
 */
export function assetBalanceFromDto(
  dto: TgvalidatordAssetBalance | null | undefined
): AssetBalance | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const info = dto.asset?.currencyInfo;
  return {
    currencyId: info?.id,
    currency: info?.symbol ?? dto.asset?.currency,
    blockchain: info?.blockchain,
    network: info?.network,
    contractAddress: info?.contractAddress,
    tokenId: dto.asset?.nft?.tokenid ?? info?.tokenID,
    balance: dto.balance?.totalConfirmed,
    fiatValue: undefined,
    fiatCurrency: undefined,
  };
}

/**
 * Maps an array of balance rows to AssetBalance domain models.
 */
export function assetBalancesFromDto(
  dtos: TgvalidatordAssetBalance[] | null | undefined
): AssetBalance[] {
  return safeMap(dtos, assetBalanceFromDto);
}

/**
 * Maps an NFT collection balance row to an NFTCollectionBalance domain model.
 *
 * The row is `{currencyInfo, balance}`: the collection is described by its currency and
 * the count is the confirmed balance.
 */
export function nftCollectionBalanceFromDto(
  dto: TgvalidatordNFTCollectionBalance | null | undefined
): NFTCollectionBalance | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const info = dto.currencyInfo;
  return {
    name: info?.name,
    symbol: info?.symbol,
    blockchain: info?.blockchain,
    network: info?.network,
    contractAddress: info?.contractAddress,
    count: safeInt(dto.balance?.totalConfirmed),
    logoUrl: info?.logo,
  };
}

/**
 * Maps an array of NFT collection balance rows to NFTCollectionBalance domain models.
 */
export function nftCollectionBalancesFromDto(
  dtos: TgvalidatordNFTCollectionBalance[] | null | undefined
): NFTCollectionBalance[] {
  return safeMap(dtos, nftCollectionBalanceFromDto);
}

File: taurus-protect-sdk-go/pkg/protect/mapper/balance.go
package mapper

import (
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// AssetBalanceFromDTO converts an OpenAPI AssetBalance to a domain AssetBalance.
func AssetBalanceFromDTO(dto *openapi.TgvalidatordAssetBalance) *model.AssetBalance {
	if dto == nil {
		return nil
	}

	return &model.AssetBalance{
		Asset:   AssetFromDTO(dto.Asset),
		Balance: BalanceFromDTO(dto.Balance),
	}
}

// AssetBalancesFromDTO converts a slice of OpenAPI AssetBalance to domain AssetBalances.
func AssetBalancesFromDTO(dtos []openapi.TgvalidatordAssetBalance) []*model.AssetBalance {
	if dtos == nil {
		return nil
	}
	balances := make([]*model.AssetBalance, len(dtos))
	for i := range dtos {
		balances[i] = AssetBalanceFromDTO(&dtos[i])
	}
	return balances
}

// AssetFromDTO converts an OpenAPI Asset to a domain Asset.
func AssetFromDTO(dto *openapi.TgvalidatordAsset) *model.Asset {
	if dto == nil {
		return nil
	}

	return &model.Asset{
		Currency:     dto.Currency,
		Kind:         safeString(dto.Kind),
		CurrencyInfo: CurrencyInfoFromDTO(dto.CurrencyInfo),
	}
}

// CurrencyInfoFromDTO converts an OpenAPI Currency to a domain CurrencyInfo.
func CurrencyInfoFromDTO(dto *openapi.TgvalidatordCurrency) *model.CurrencyInfo {
	if dto == nil {
		return nil
	}

	return &model.CurrencyInfo{
		ID:              safeString(dto.Id),
		Name:            safeString(dto.Name),
		Symbol:          safeString(dto.Symbol),
		DisplayName:     safeString(dto.DisplayName),
		Blockchain:      safeString(dto.Blockchain),
		Network:         safeString(dto.Network),
		Decimals:        safeString(dto.Decimals),
		ContractAddress: safeString(dto.ContractAddress),
		TokenID:         safeString(dto.TokenID),
		Type:            safeString(dto.Type),
		IsToken:         safeBool(dto.IsToken),
		IsERC20:         safeBool(dto.IsERC20),
		IsNFT:           safeBool(dto.IsNFT),
		IsFiat:          safeBool(dto.IsFiat),
		Enabled:         safeBool(dto.Enabled),
	}
}
